### PR TITLE
PLT-516 Part 1 of performance fixes for large teams

### PR DESCRIPTION
**Pull request by jwilander**
##### This includes three major performance improvements for large teams:
1. `getProfiles` no longer does a bunch of slow event emitting/json serializing for every received profile
2. `getStatuses` now only returns the user statuses for the list of provided user ids instead of a status for every user on the team
3. The LHS no longer loops through all user profiles to populate direct messages but rather checks existing channels properly (possible due to new direct channels modal)

##### Some stuff coming in part 2 tomorrow:
1. More intelligently load user profile pictures
2. Return less data in `getProfiles`

##### Related tasks to do for 1.3:
1. Return less data in all API services where possible
2. Rework how user statuses are handled
3. Look in to more refactoring of LHS
4. Handling user profiles better

### CODE REVIEW

**Comment by jwilander** [comment]
Should help with #888

File: web/react/stores/user_store.jsx [hunk 1]
```diff
@@ -235,6 +237,14 @@ class UserStoreClass extends EventEmitter {
         BrowserStore.setItem('profiles', ps);
     }
 
+    saveProfiles(profiles) {
+        const currentId = this.getCurrentId();
+        if (currentId in profiles) {
+            delete profiles[currentId];
```
Comment (coreyhulen) [inline web/react/stores/user_store.jsx]: Should you just save over this? vs deleting? just nervous that other places may expect current user to be in this list 0/5.

**Comment by coreyhulen** [comment]
+1

File: web/react/stores/user_store.jsx [hunk 2]
```diff
@@ -320,15 +330,8 @@ UserStore.dispatchToken = AppDispatcher.register((payload) => {
 
     switch (action.type) {
     case ActionTypes.RECIEVED_PROFILES:
-        for (var id in action.profiles) {
-            // profiles can have incomplete data, so don't overwrite current user
-            if (id === UserStore.getCurrentId()) {
```
Comment (coreyhulen) [inline web/react/stores/user_store.jsx]: ah..nvm about my comment above.

File: web/react/components/sidebar.jsx [hunk 3]
```diff
@@ -100,74 +100,56 @@ export default class Sidebar extends React.Component {
...
+        const currentChannelId = ChannelStore.getCurrentId();
 
-        var teammates = [];
-        for (var id in teamMemberMap) {
-            if (id === currentUserId) {
-                continue;
-            }
-            teammates.push(teamMemberMap[id]);
-        }
+        const channels = Object.assign([], ChannelStore.getAll());
```
Comment (hmhealey) [inline web/react/components/sidebar.jsx]: `channels` is never mutated so copying it seems unnecessary. I'm also not sure if Object.assign actually works for arrays, but I could be wrong

**Comment by hmhealey** [comment]
Other than that one small thing I pointed out, everything else looks good

File: web/react/components/sidebar.jsx (hunk 3 above)
Comment (jwilander) [inline web/react/components/sidebar.jsx]: It works for arrays and yeah I guess there's no need to copy in this case

Comment (hmhealey) [inline web/react/components/sidebar.jsx]: Cool, sounds good

**Pull request merged**